src/function/macro_function.cpp

```diff
@@ -0,0 +1,56 @@
+#include "duckdb/function/macro_function.hpp"
+
+#include "duckdb/catalog/catalog_entry/macro_catalog_entry.hpp"
+#include "duckdb/catalog/catalog_entry/scalar_function_catalog_entry.hpp"
+#include "duckdb/parser/expression/columnref_expression.hpp"
+#include "duckdb/parser/expression/function_expression.hpp"
+
+namespace duckdb {
+
+MacroFunction::MacroFunction(unique_ptr<ParsedExpression> expression) : expression(move(expression)) {
+}
+
+string MacroFunction::ValidateArguments(ClientContext &context, QueryErrorContext &error_context,
+                                        MacroCatalogEntry &macro_func, FunctionExpression &function_expr) {
+	string error;
+	auto &catalog = Catalog::GetCatalog(context);
+	auto &parameters = macro_func.function->parameters;
+	auto &arguments = function_expr.children;
+	if (parameters.size() != arguments.size()) {
+		error = StringUtil::Format(
+		    "Macro function '%s(%s)' requires ", macro_func.name,
+		    StringUtil::Join(parameters, parameters.size(), ", ", [](const unique_ptr<ParsedExpression> &p) {
+			    return ((ColumnRefExpression &)*p).column_name;
+		    }));
+		error += parameters.size() == 1 ? "a single argument" : StringUtil::Format("%i arguments", parameters.size());
+		error += ", but ";
+		error +=
+		    arguments.size() == 1 ? "a single argument was" : StringUtil::Format("%i arguments were", arguments.size());
+		error += " provided.";
+		return error;
+	}
+
+	// check for arguments with side-effects TODO: to support this, a projection must be pushed
+	for (auto &arg : arguments) {
+		if (arg->GetExpressionClass() == ExpressionClass::FUNCTION) {
```

Comment:
This is actually not entirely correct, since an argument could have hidden side effects (e.g. `i + nextval('seq')`). My CSE optimizer branch (#1143) has a fix for this by adding the `HasSideEffects()` function to expressions, which can be used here. I will merge that one first. Could you then change the side effect detection here to use that function and add a testcase?

Reply:
The loop iterates over ParsedExpressions, not Expressions, so the side-effects trick will not work here. I could try to call BindChild to transform it into an Expression, and then call HasSideEffects, not sure if that will work. Will give it a go